src/Coordination/KeeperStorage.cpp

```diff
@@ -198,356 +176,650 @@ void KeeperStorage::Node::removeChild(StringRef child_path)
 }
 
 KeeperStorage::KeeperStorage(int64_t tick_time_ms, const String & superdigest_)
-    : session_expiry_queue(tick_time_ms)
-    , superdigest(superdigest_)
+    : session_expiry_queue(tick_time_ms), superdigest(superdigest_)
 {
     container.insert("/", Node());
 }
 
-using Undo = std::function<void()>;
+template <class... Ts>
+struct Overloaded : Ts...
+{
+    using Ts::operator()...;
+};
+template <class... Ts>
+Overloaded(Ts...) -> Overloaded<Ts...>;
+
+std::shared_ptr<KeeperStorage::Node> KeeperStorage::UncommittedState::getNode(StringRef path)
+{
+    std::shared_ptr<Node> node{nullptr};
+
+    if (auto maybe_node_it = storage.container.find(path); maybe_node_it != storage.container.end())
+    {
+        const auto & committed_node = maybe_node_it->value;
+        node = std::make_shared<KeeperStorage::Node>();
+        node->stat = committed_node.stat;
+        node->seq_num = committed_node.seq_num;
+        node->setData(committed_node.getData());
+    }
+
+    applyDeltas(
+        path,
+        Overloaded{
+            [&](const CreateNodeDelta & create_delta)
+            {
+                assert(!node);
+                node = std::make_shared<Node>();
+                node->stat = create_delta.stat;
+                node->setData(create_delta.data);
+            },
+            [&](const RemoveNodeDelta & /*remove_delta*/)
+            {
+                assert(node);
+                node = nullptr;
+            },
+            [&](const UpdateNodeDelta & update_delta)
+            {
+                assert(node);
+                update_delta.update_fn(*node);
+            },
+            [&](auto && /*delta*/) {},
+        });
+
+    return node;
```

Comment:
What does `nullptr` result mean?

Reply:
Node doesn't exist (if it's present in storage than it's deleted by RemoveNodeDelta)